File: research/chr-symbolic/matching.py
"""Read-only finite-tree views for symbolic head matching and equality guards.

Datatype expressions are determined by the heap. Free source variables are Hole
constructors with explicit IDs, not unconstrained solver-selected ground terms.
"""
import z3
from heap import select

_SORTS={}


def tree_sort(signature):
    key=tuple(signature)
    if key not in _SORTS:
        name=f'CHRTree{len(_SORTS)}'
        declaration=z3.Datatype(name)
        declaration.declare(f'{name}_hole',(f'{name}_id',z3.IntSort()))
        for i,(_,arity) in enumerate(signature):
            declaration.declare(f'{name}_c{i}',*[(f'{name}_c{i}_{j}',declaration) for j in range(arity)])
        _SORTS[key]=declaration.create()
    return _SORTS[key]


class View:
    def __init__(self,heap):
        self.h=heap;self.sort=tree_sort(heap.signature)
        self.hole=self.sort.constructor(0)
        self.next_var=heap.variables
        values=[self.hole(i) for i in heap.ids]
        # A valid allocated heap has no cycle after combining constructor and
        # substitution edges. n rounds cover every path from an allocated root.
        for _ in range(heap.n):
            updated=[]
            for i in range(heap.n):
                value=self.hole(heap.ids[i])
                for tag,(_,arity) in enumerate(heap.signature):
                    app=self.sort.constructor(tag+1)(*[select(values,heap.children[k][i]) for k in range(arity)])
                    value=z3.If(heap.tags[i]==tag+2,app,value)
                updated.append(heap.e.bind(z3.If(heap.sub[i]!=i,select(values,heap.sub[i]),value)))
            values=updated
        self.values=values

    def match(self,pattern,handle,bindings=None):
        bindings={} if bindings is None else dict(bindings)
        def visit(p,value):
            value=self.h.deref(value)
            if isinstance(p,int):
                if p in bindings:return select(self.values,bindings[p])==select(self.values,value)
                bindings[p]=value
                return z3.BoolVal(True)
            name,args=p
            if (name,len(args)) not in self.h.signature:return z3.BoolVal(False)
            tag=self.h.signature.index((name,len(args)))+2
            return z3.And(select(self.h.tags,value)==tag,*[visit(child,select(self.h.children[k],value)) for k,child in enumerate(args)])
        return self.h.e.bind(visit(pattern,handle)),bindings

    def guard(self,left,right,bindings):
        local={v:select(self.values,handle) for v,handle in bindings.items()}
        fresh=0
        def instantiate(t):
            nonlocal fresh
            if isinstance(t,int):
                if t not in local:
                    local[t]=self.hole(self.next_var+fresh);fresh+=1
                return local[t]
            name,args=t
            if (name,len(args)) not in self.h.signature:raise ValueError('guard constructor outside signature')
            index=self.h.signature.index((name,len(args)))+1
            return self.sort.constructor(index)(*[instantiate(c) for c in args])
        return self.h.e.bind(instantiate(left)==instantiate(right))


from dataclasses import dataclass
from itertools import permutations


@dataclass
class Application:
    rule: int
    occurrences: tuple
    bindings: dict
    selected: object


def applications(heap,rules,roots,alive,history):
    """One-hot committed selection over fixed occurrence slots.

    Slot indices are monotone occurrence identities. Inactive slots cannot match;
    permutations exclude using the same occurrence twice. This emits factored
    enabling predicates, not a table of concrete reachable stores.
    """
    if len(roots)!=len(alive):raise ValueError('occurrence vectors differ')
    view=View(heap)
    found=z3.BoolVal(False)
    result=[]
    for ri,rule in enumerate(rules):
        heads=rule.kept+rule.removed
        if not heads:raise ValueError('empty head')
        for ids in permutations(range(len(roots)),len(heads)):
            bindings={}
            enabled=[alive[i] for i in ids]
            enabled.append(z3.Not(history.get((ri,ids),z3.BoolVal(False))))
            for head,oi in zip(heads,ids):
                ok,bindings=view.match(head,roots[oi],bindings)
                enabled.append(ok)
            enabled.extend(view.guard(a,b,bindings) for a,b in rule.guards)
            ready=heap.e.bind(z3.And(*enabled))
            selected=heap.e.bind(z3.And(z3.Not(found),ready))
            result.append(Application(ri,ids,bindings,selected))
            found=heap.e.bind(z3.Or(found,ready))
    return result

File: docs/experiments/results/s05-call-generated-entry/registered/crates/chr-reference/src/matching.rs
//! Exhaustive, nonbinding head matching in a frozen branch.
use crate::{
    Stats,
    branch::{Branch, Token},
    instantiate::{self, Bindings},
};
use chr_syntax::{Constraint, Goal, Guard, Rule, Term};

pub(crate) struct Application {
    pub token: Token,
    pub removed: Vec<u64>,
    pub body: Goal,
    pub next_var: u64,
}

pub(crate) fn find(branch: &Branch, rules: &[Rule], stats: &mut Stats) -> Option<Application> {
    for (rule_index, rule) in rules.iter().enumerate() {
        let heads = rule.kept.iter().chain(&rule.removed).collect::<Vec<_>>();
        let matcher = Matcher {
            branch,
            rule,
            rule_index,
            heads,
        };
        if let Some(result) = matcher.search(&mut vec![], &Bindings::new(), stats) {
            return Some(result);
        }
    }
    None
}

struct Matcher<'a> {
    branch: &'a Branch,
    rule: &'a Rule,
    rule_index: usize,
    heads: Vec<&'a Constraint>,
}
impl Matcher<'_> {
    fn search(
        &self,
        selected: &mut Vec<usize>,
        bindings: &Bindings,
        stats: &mut Stats,
    ) -> Option<Application> {
        if selected.len() == self.heads.len() {
            let ids = selected
                .iter()
                .map(|i| self.branch.store[*i].id)
                .collect::<Vec<_>>();
            let token = (self.rule_index, ids.clone());
            if self.branch.history.contains(&token) {
                return None;
            }
            let mut bindings = bindings.clone();
            let mut next_var = self.branch.next_var;
            for Guard::Equal(a, b) in &self.rule.guards {
                let a = instantiate::term(a, &mut bindings, &mut next_var);
                let b = instantiate::term(b, &mut bindings, &mut next_var);
                if self.branch.substitution.resolve(&a) != self.branch.substitution.resolve(&b) {
                    return None;
                }
            }
            let body = instantiate::goal(&self.rule.body, &mut bindings, &mut next_var);
            return Some(Application {
                token,
                removed: ids[self.rule.kept.len()..].to_vec(),
                body,
                next_var,
            });
        }
        let head = self.heads[selected.len()];
        for (index, occurrence) in self.branch.store.iter().enumerate() {
            if selected.contains(&index) {
                continue;
            }
            let value = &occurrence.constraint;
            if head.name != value.name || head.args.len() != value.args.len() {
                continue;
            }
            stats.head_candidates += 1;
            let mut extended = bindings.clone();
            if head
                .args
                .iter()
                .zip(&value.args)
                .all(|(p, t)| match_term(p, &self.branch.substitution.resolve(t), &mut extended))
            {
                selected.push(index);
                let result = self.search(selected, &extended, stats);
                selected.pop();
                if result.is_some() {
                    return result;
                }
            }
        }
        None
    }
}

fn match_term(pattern: &Term, value: &Term, bindings: &mut Bindings) -> bool {
    match pattern {
        Term::Var(var) => match bindings.get(var) {
            Some(previous) => previous == value,
            None => {
                bindings.insert(*var, value.clone());
                true
            }
        },
        Term::App(f, xs) => match value {
            Term::App(g, ys) if f == g && xs.len() == ys.len() => {
                xs.iter().zip(ys).all(|(x, y)| match_term(x, y, bindings))
            }
            _ => false,
        },
    }
}
